src/rules/sort-collections.ts

```diff
@@ -1,125 +1,157 @@
 
-const defaultCollections = [
-	"scripts",
-	"devDependencies",
-	"dependencies",
-	"peerDependencies",
-	"overrides",
+const defaultCollections = new Set([
 	"config",
+	"dependencies",
+	"dependenciesMeta",
+	"devDependencies",
 	"exports",
-];
+	"optionalDependencies",
+	"overrides",
+	"peerDependencies",
+	"peerDependenciesMeta",
+	"pnpm.allowedDeprecatedVersions",
+	"pnpm.overrides",
+	"pnpm.packageExtensions",
+	"pnpm.patchedDependencies",
+	"pnpm.peerDependencyRules.allowedVersions",
+	"resolutions",
+	"scripts",
+]);
 
 type Options = string[];
 
 export const rule = createRule<Options>({
 	create(context) {
-		const toSort = context.options[0] || defaultCollections;
+		const toSort = context.options[0]
+			? new Set(context.options[0])
+			: defaultCollections;
+
 		return {
 			"JSONProperty:exit"(node) {
-				const { key, value } = node as AST.JSONProperty & {
-					key: AST.JSONStringLiteral;
-				};
+				const { key: nodeKey, value: collection } = node;
 
-				const collection = value;
 				if (
-					collection.type === "JSONObjectExpression" &&
-					toSort.includes(key.value)
+					nodeKey.type !== "JSONLiteral" ||
+					collection.type !== "JSONObjectExpression"
+				) {
+					return;
+				}
+
+				const keyPartsReversed = [nodeKey.value];
+				for (
+					let currNode: JsonAST.JSONNode | null | undefined =
+						node.parent;
+					// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+					currNode;
+					currNode = currNode.parent
 				) {
-					const currentOrder = collection.properties;
-					const scripts = new Set(
-						currentOrder.map(
-							(prop) => (prop.key as AST.JSONStringLiteral).value,
-						),
-					);
-
-					const desiredOrder = currentOrder.slice().sort((a, b) => {
-						let aKey = (a.key as AST.JSONStringLiteral).value;
-						let bKey = (b.key as AST.JSONStringLiteral).value;
-
-						if (key.value !== "scripts") {
-							return aKey > bKey ? 1 : -1;
-						} else {
-							let modifier = 0;
-
-							if (
-								aKey.startsWith("pre") &&
-								scripts.has(aKey.substring(3))
-							) {
-								aKey = aKey.substring(3);
-								modifier -= 1;
-							} else if (
-								aKey.startsWith("post") &&
-								scripts.has(aKey.substring(4))
-							) {
-								aKey = aKey.substring(4);
-								modifier += 1;
-							}
-
-							if (
-								bKey.startsWith("pre") &&
-								scripts.has(bKey.substring(3))
-							) {
-								bKey = bKey.substring(3);
-								modifier += 1;
-							} else if (
-								bKey.startsWith("post") &&
-								scripts.has(bKey.substring(4))
-							) {
-								bKey = bKey.substring(4);
-								modifier -= 1;
-							}
-
-							if (aKey === bKey) {
-								return modifier;
-							}
-
-							return aKey > bKey ? 1 : -1;
-						}
-					});
 					if (
-						currentOrder.some(
-							(property, i) => desiredOrder[i] !== property,
-						)
+						currNode.type === "JSONProperty" &&
+						currNode.key.type === "JSONLiteral"
 					) {
```

Comment:
[Bug] What should happen if parent isn't a JSONProperty with a JSONLiteral key? Example:

```js
// eslint.config.js
"sort-collections": ["error", ["gotcha.value"]],
```

```jsonc
// package.json
{
  "gotcha": [
    {
      "value": 123
    }
  ]
}
```

This might actually need a redesign of the options. How do we handle this? Should it be supported at all?

My initial tentative vote: don't support it at first, and consider it an odd edge case? Then if someone files an issue we can always expand the format / add a new allowed one.

Reply:
Currently, the "full" key of the `value` property is going to be `gotcha.value`, as if there are no extra array(s) around `{"value": 123}`.

Should I keep this behavior (which seems fine to me) and if yes, then should I add a corresponding test? This indeed looks like a rare case, and I think the majority of the rule's users won't be affected by that.

Reply:
Coming back to this, I still am leaning somewhat strongly towards not keeping the behavior. Agreed it's kind of an unusual/weird case that folks aren't likely to hit. But I've seen some _funky_ package.jsons. I think it'd be better to play it safe.

cc @michaelfaith - I could be persuaded if multiple people think it's fine as-is 🙂

Reply:
I agree. It feels odd to apply this to array items.

Reply:
I've removed this behavior, now objects within arrays are unreachable with dot notation, and added corresponding test.